store/tikv/txn_scheduler.go

```diff
@@ -0,0 +1,189 @@
+	t.RLock()
+	defer t.RUnlock()
+	return t.txns[startTS]
+}
+
+func (t *txnMap) delete(startTS uint64) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.txns, startTS)
+}
+
+// txnScheduler is the scheduler to commit transactions.
+type txnScheduler struct {
+	latches *latch.Latches
+	txnMaps []txnMap
+}
+
+const (
+	txnSchedulerLatchesSize = 102400
+	txnSchedulerMapsCount   = 256
+)
+
+func newTxnScheduler(enableLatches bool) *txnScheduler {
+	if !enableLatches {
+		return &txnScheduler{
+			latches: nil,
+			txnMaps: nil,
+		}
+	}
+	return &txnScheduler{
+		latches: latch.NewLatches(txnSchedulerLatchesSize),
+		txnMaps: make([]txnMap, txnSchedulerMapsCount),
+	}
+}
+
+func (scheduler *txnScheduler) execute(ctx context.Context, txn *twoPhaseCommitter, connID uint64) error {
+	newCommitter := newTxnCommiter(ctx, txn)
+	// latches disabled
+	if scheduler.latches == nil {
+		_, err := newCommitter.execute()
+		return errors.Trace(err)
+	}
+	// latches enabled
+	newCommitter.lock = scheduler.latches.GenLock(txn.startTS, txn.keys)
+	// for transactions which is not retryable, commit directly.
+	if !sessionctx.GetRetryAble(ctx) {
+		err := scheduler.runForUnRetryAble(newCommitter)
+		return errors.Trace(err)
+	}
+	// for transactions which need to acquire latches
+	ch := make(chan error, 1)
+	newCommitter.ch = ch
+	scheduler.putTxn(txn.startTS, newCommitter)
+	go scheduler.run(txn.startTS)
+
+	err := errors.Trace(<-ch)
+	return errors.Trace(err)
+}
+
+func (scheduler *txnScheduler) getMap(startTS uint64) *txnMap {
+	id := int(startTS % uint64(len(scheduler.txnMaps)))
+	return &scheduler.txnMaps[id]
+}
+
+func (scheduler *txnScheduler) getTxn(startTS uint64) *txnCommitter {
+	return scheduler.getMap(startTS).get(startTS)
+}
+
+func (scheduler *txnScheduler) deleteTxn(startTS uint64) {
+	scheduler.getMap(startTS).delete(startTS)
+}
+
+func (scheduler *txnScheduler) putTxn(startTS uint64, txn *txnCommitter) {
+	scheduler.getMap(startTS).put(startTS, txn)
+}
+
+// run the transaction directly which is not retryable.
+func (scheduler *txnScheduler) runForUnRetryAble(txn *txnCommitter) error {
+	commitTS, err := txn.execute()
+	if err == nil {
+		go scheduler.latches.RefreshCommitTS(txn.lock.RequiredSlots(), commitTS)
+	}
+	log.Debug(txn.twoPC.connID, " finish txn with startTS:", txn.twoPC.startTS, " commitTS:", commitTS, " error:", err)
+	return errors.Trace(err)
+}
+
+// run a normal transaction, it needs to acquire latches before 2pc.
+func (scheduler *txnScheduler) run(startTS uint64) {
+	txn := scheduler.getTxn(startTS)
+	if txn == nil {
+		log.Error("Transaction with StartTS:", startTS, " is not exist")
+		return
+	}
+	acquired, stale := scheduler.latches.Acquire(txn.lock)
+	if !stale && !acquired {
+		// wait for next wakeup
+		return
+	}
+	commitTS := txn.commit(stale)
+	scheduler.deleteTxn(startTS)
+	wakeupList := scheduler.latches.Release(txn.lock, commitTS)
+	for _, s := range wakeupList {
+		scheduler.run(s)
```

Comment:
We should run wakeup transactions in background goroutine, but run directly at line 138